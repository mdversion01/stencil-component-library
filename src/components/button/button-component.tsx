import { Component, Prop, Event, EventEmitter, h } from '@stencil/core';
import { logInfo } from '../../utils/log-debug';

@Component({
  tag: 'button-component',
  styleUrl: 'button.scss',
  // styleUrls: ['../../global/global.scss',],
})
export class Button {
  @Prop() absolute: boolean = false;
  @Prop() active: boolean = false;
  @Prop() ariaLabel: string = '';
  @Prop() block: boolean = false;
  @Prop() bottom: string = '';
  @Prop() btnIcon: boolean = false;
  @Prop() btnText: string = '';
  @Prop() classNames: string = '';
  @Prop() disabled: boolean = false;
  @Prop() end: boolean = false;
  @Prop() elevation: string = '';
  @Prop() fixed: boolean = false;
  @Prop() groupBtn: boolean = false;
  @Prop() iconBtn: boolean = false;
  @Prop({ reflect: true }) slotSide: 'left' | 'right' | 'none' = 'none';
  @Prop() styles: string = '';
  @Prop() left: string = '';
  @Prop() link: boolean = false;
  @Prop() outlined: boolean = false;
  @Prop() pressed: boolean | string = false;
  @Prop() right: string = '';
  @Prop() ripple: boolean = false;
  @Prop() shape: string = '';
  @Prop() size: string = '';
  @Prop() start: boolean = false;
  @Prop() stripped: boolean = false;
  @Prop() text: boolean = false;
  @Prop() textBtn: boolean = false;
  @Prop() titleAttr: string = '';
  @Prop() top: string = '';
  @Prop() url: string = '';
  @Prop() variant: string = '';
  @Prop() vertical: boolean = false;
  @Prop() zIndex: string = '';
  @Prop({ mutable: true }) isOpen: boolean = false;
  @Prop() targetId: string = '';
  @Prop() accordion: boolean = false;
  @Prop() devMode: boolean = false;

  @Event() customClick: EventEmitter;

  private handleClick() {
    this.customClick.emit();
    logInfo(this.devMode, 'Button', 'Clicked');
  }

  private handleKeyDown = (event: KeyboardEvent) => {
    if ((event.key === 'Enter' || event.key === ' ') && !this.disabled) {
      event.preventDefault();
      this.handleClick();
    }
  };

  private getDynamicStyles(): { [key: string]: string } {
    const styles: { [key: string]: string } = {};
    if (this.absolute) styles.position = 'absolute';
    if (this.fixed) styles.position = 'fixed';
    if (this.left) styles.left = `${this.left}px`;
    if (this.right) styles.right = `${this.right}px`;
    if (this.top) styles.top = `${this.top}px`;
    if (this.bottom) styles.bottom = `${this.bottom}px`;
    if (this.zIndex) styles['z-index'] = this.zIndex;

    logInfo(this.devMode, 'Button', 'getDynamicStyles()', styles);
    return styles;
  }

  private renderButtonContent() {
    const hasLeft = this.slotSide === 'left';
    const hasRight = this.slotSide === 'right';

    const content = this.btnIcon ? (
      <slot />
    ) : this.iconBtn ? (
      <span class="btn__content">
        <slot />
      </span>
    ) : (
      [
        hasLeft && <slot />,
        <span class={`btn__content ${hasLeft ? 'ms-1' : ''} ${hasRight ? 'me-1' : ''}`}>
          {this.btnText}
          {!hasLeft && !hasRight && <slot />}
        </span>,
        hasRight && <slot />,
      ]
    );

    logInfo(this.devMode, 'Button', 'renderButtonContent()', {
      btnText: this.btnText,
      slotSide: this.slotSide,
      btnIcon: this.btnIcon,
      iconBtn: this.iconBtn,
    });

    return content;
  }

  render() {
    const dynamicStyles = this.getDynamicStyles();
    const isGroup = !!this.groupBtn;
    const buttonGroup = this.vertical ? 'btn-group-vertical' : 'btn-group';
    const placement = this.start ? `${buttonGroup}-start` : this.end ? `${buttonGroup}-end` : 'btn-group__btn';

    const classList = [
      'btn',
      this.classNames,
      this.outlined ? `btn-outline-${this.variant}` : '',
      !this.outlined && this.variant ? `btn-${this.variant}` : '',
      this.block && 'btn--block',
      this.size && `btn-${this.size}`,
      this.shape,
      this.active && 'active',
      isGroup && buttonGroup,
      isGroup && placement,
      this.ripple && 'btn-ripple',
      this.stripped && 'stripped',
      this.iconBtn && 'icon-btn',
      this.link && 'link',
      this.elevation && `elevated-${this.elevation}`,
      this.textBtn && 'text-btn',
      this.text && 'text',
      this.btnIcon && 'btn-icon',
    ]
      .filter(Boolean)
      .join(' ');

    // Accessibility: ensure an aria-label when there is no visible text
    const hasVisibleText = !!(this.btnText && this.btnText.trim().length);
    const computedAriaLabel =
      (this.ariaLabel && this.ariaLabel.trim()) ||
      (hasVisibleText ? this.btnText.trim() : '') ||
      (this.titleAttr && this.titleAttr.trim()) ||
      'Button';

    if (!hasVisibleText && !this.ariaLabel && (this.btnIcon || this.iconBtn) && this.devMode) {
      console.warn(
        '[ButtonComponent] Icon-only button has no `ariaLabel`/text. Setting aria-label="%s".',
        computedAriaLabel
      );
    }

    logInfo(this.devMode, 'Button', 'render()', {
      classList,
      link: this.link,
      disabled: this.disabled,
      slotSide: this.slotSide,
      computedAriaLabel,
    });

    const ariaProps: Record<string, any> = {};
    if (this.accordion && this.targetId) {
      ariaProps['aria-expanded'] = String(this.isOpen);
      ariaProps['aria-controls'] = this.targetId;
    }

    const buttonProps = {
      ...ariaProps,
      'aria-pressed': this.pressed === 'true' || this.pressed === 'false' ? this.pressed : undefined,
      'aria-label': computedAriaLabel,
      'aria-disabled': this.link && this.disabled ? 'true' : undefined,
      style: dynamicStyles,
      class: classList,
      disabled: !this.link && this.disabled,
      onClick: (event: MouseEvent) => {
        if (this.disabled) {
          event.preventDefault();
          return;
        }
        const shouldPrevent = !this.url || this.url === '#' || this.url.trim() === '' || this.url.startsWith('#');
        if (this.link && shouldPrevent) event.preventDefault();
        this.handleClick();
      },
    };

    return this.link ? (
      <a
        {...(buttonProps as any)}
        href={this.url && this.url.trim() !== '' ? this.url : undefined}
        role="button"
        tabindex={this.disabled ? '-1' : '0'}
        onKeyDown={this.disabled ? undefined : this.handleKeyDown}
        title={this.titleAttr}
      >
        {this.renderButtonContent()}
      </a>
    ) : (
      <button {...(buttonProps as any)} type="button" title={this.titleAttr}>
        {this.renderButtonContent()}
      </button>
    );
  }
}
