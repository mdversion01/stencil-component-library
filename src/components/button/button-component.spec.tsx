// src/components/button/button-component.new.spec.tsx
import { newSpecPage } from '@stencil/core/testing';
import { Button } from './button-component';

const getControl = (root: HTMLElement) =>
  (root.querySelector('button') as HTMLElement) || (root.querySelector('a') as HTMLElement);

describe('button-component (new spec)', () => {
  it('renders default <button> with text and stable aria-label', async () => {
    const page = await newSpecPage({
      components: [Button],
      html: `<button-component btn-text="Click me"></button-component>`,
    });
    const btn = getControl(page.root);
    expect(page.root.getAttribute('slot-side')).toBe('none');
    expect(btn).toBeTruthy();
    expect(btn.tagName).toBe('BUTTON');
    expect(btn.getAttribute('aria-label')).toBe('Click me');
    expect(btn).toMatchSnapshot();
  });

  it('renders <a> link with aria-disabled and proper role/tabindex when disabled', async () => {
    const page = await newSpecPage({
      components: [Button],
      html: `<button-component link disabled btn-text="Disabled Link"></button-component>`,
    });
    const a = getControl(page.root);
    expect(a.tagName).toBe('A');
    expect(a.getAttribute('aria-disabled')).toBe('true');
    expect(a.getAttribute('role')).toBe('button');
    expect(a.getAttribute('tabindex')).toBe('-1');
    expect(a).toMatchSnapshot();
  });

  it('applies ARIA attributes (aria-pressed, aria-label) on native button', async () => {
    const page = await newSpecPage({
      components: [Button],
      html: `<button-component btn-text="Submit" aria-label="Submit form" pressed="true"></button-component>`,
    });
    const btn = getControl(page.root);
    expect(btn.getAttribute('aria-pressed')).toBe('true');
    expect(btn.getAttribute('aria-label')).toBe('Submit form');
    expect(btn).toMatchSnapshot();
  });

  it('slotSide="left" adds ms-1 and keeps me-1 off', async () => {
    const page = await newSpecPage({
      components: [Button],
      html: `<button-component btn-text="With Icon" slot-side="left"><span>üåü</span></button-component>`,
    });
    const content = page.root.querySelector('.btn__content') as HTMLElement;
    expect(page.root.getAttribute('slot-side')).toBe('left');
    expect(content.classList.contains('ms-1')).toBe(true);
    expect(content.classList.contains('me-1')).toBe(false);
    expect(getControl(page.root)).toMatchSnapshot();
  });

  it('slotSide="right" adds me-1 and keeps ms-1 off', async () => {
    const page = await newSpecPage({
      components: [Button],
      html: `<button-component btn-text="Next" slot-side="right"><span>‚û°Ô∏è</span></button-component>`,
    });
    const content = page.root.querySelector('.btn__content') as HTMLElement;
    expect(page.root.getAttribute('slot-side')).toBe('right');
    expect(content.classList.contains('me-1')).toBe(true);
    expect(content.classList.contains('ms-1')).toBe(false);
    expect(getControl(page.root)).toMatchSnapshot();
  });

  it('btnIcon (icon-only) renders slot directly, with aria-label fallback + dev warning', async () => {
    const warn = jest.spyOn(console, 'warn').mockImplementation(() => {});
    const page = await newSpecPage({
      components: [Button],
      html: `<button-component btn-icon dev-mode><span>üîî</span></button-component>`,
    });
    const btn = getControl(page.root);
    expect(btn.getAttribute('aria-label')).toBe('Button'); // fallback
    expect(warn).toHaveBeenCalled();
    warn.mockRestore();
    expect(btn).toMatchSnapshot();
  });

  it('iconBtn wraps slot in .btn__content', async () => {
    const page = await newSpecPage({
      components: [Button],
      html: `<button-component icon-btn><span>üîî</span></button-component>`,
    });
    expect(page.root.querySelector('.btn__content')).toBeTruthy();
    expect(getControl(page.root)).toMatchSnapshot();
  });

  it('grouped button with start placement adds group classes', async () => {
    const page = await newSpecPage({
      components: [Button],
      html: `<button-component group-btn start btn-text="Start Btn"></button-component>`,
    });
    const btn = getControl(page.root);
    expect(btn.classList.contains('btn-group')).toBe(true);
    expect(btn.classList.contains('btn-group-start')).toBe(true);
    expect(btn).toMatchSnapshot();
  });

  it('applies variant/size/elevation classes (presence only)', async () => {
    const page = await newSpecPage({
      components: [Button],
      html: `<button-component btn-text="Elevated" elevation="2" variant="primary" size="lg"></button-component>`,
    });
    const btn = getControl(page.root);
    expect(btn.classList.contains('elevated-2')).toBe(true);
    expect(btn.classList.contains('btn-primary')).toBe(true);
    expect(btn.classList.contains('btn-lg')).toBe(true);
  });

  it('keyboard Enter/Space on link emits customClick (when enabled)', async () => {
    const page = await newSpecPage({
      components: [Button],
      html: `<button-component link btn-text="Press Me" url="#"></button-component>`,
    });
    const a = getControl(page.root);
    const spy = jest.fn();
    page.root.addEventListener('customClick', spy);

    a.dispatchEvent(new KeyboardEvent('keydown', { key: 'Enter', bubbles: true }));
    await page.waitForChanges();
    a.dispatchEvent(new KeyboardEvent('keydown', { key: ' ', bubbles: true }));
    await page.waitForChanges();

    expect(spy).toHaveBeenCalledTimes(2);
  });

  it('disabled link: Enter/Space do not emit', async () => {
    const page = await newSpecPage({
      components: [Button],
      html: `<button-component link disabled btn-text="Disabled Link" url="#disabled"></button-component>`,
    });
    const a = getControl(page.root);
    const spy = jest.fn();
    page.root.addEventListener('customClick', spy);

    a.dispatchEvent(new KeyboardEvent('keydown', { key: 'Enter', bubbles: true }));
    await page.waitForChanges();
    a.dispatchEvent(new KeyboardEvent('keydown', { key: ' ', bubbles: true }));
    await page.waitForChanges();

    expect(spy).not.toHaveBeenCalled();
  });

  it('disabled native button: click does not emit', async () => {
    const page = await newSpecPage({
      components: [Button],
      html: `<button-component disabled btn-text="Disabled Button"></button-component>`,
    });
    const btn = getControl(page.root);
    const spy = jest.fn();
    page.root.addEventListener('customClick', spy);

    btn.dispatchEvent(new MouseEvent('click', { bubbles: true }));
    await page.waitForChanges();

    expect(spy).not.toHaveBeenCalled();
  });

  it('accordion props: sets aria-expanded/controls and toggles when isOpen changes', async () => {
    const page = await newSpecPage({
      components: [Button],
      html: `<button-component accordion is-open="false" target-id="panel-1" btn-text="Toggle Section"></button-component>`,
    });
    const comp = page.rootInstance as Button;
    const btn = getControl(page.root);

    expect(btn.getAttribute('aria-expanded')).toBe('false');
    expect(btn.getAttribute('aria-controls')).toBe('panel-1');

    comp.isOpen = true; // mutable
    await page.waitForChanges();

    expect(btn.getAttribute('aria-expanded')).toBe('true');
    expect(btn).toMatchSnapshot();
  });

  it('dynamic styles: absolute positioning + offsets + zIndex reflected in style attribute', async () => {
    const page = await newSpecPage({
      components: [Button],
      html: `<button-component btn-text="Styled" absolute left="10" top="5" z-index="9"></button-component>`,
    });
    const btn = getControl(page.root);
    const style = (btn.getAttribute('style') || '').replace(/\s+/g, ' ');
    expect(style).toContain('position: absolute');
    expect(style).toContain('left: 10px');
    expect(style).toContain('top: 5px');
    expect(style).toContain('z-index: 9');
  });

  it('link path: prevents navigation when url is empty or "#" and still emits customClick', async () => {
    const prevent = jest.fn();
    const page = await newSpecPage({
      components: [Button],
      html: `<button-component link btn-text="NoNav" url="#"></button-component>`,
    });
    const a = getControl(page.root);
    const clickSpy = jest.fn();
    page.root.addEventListener('customClick', clickSpy);

    // simulate click with preventDefault spy
    const evt = new MouseEvent('click', { bubbles: true, cancelable: true }) as any;
    evt.preventDefault = prevent;
    a.dispatchEvent(evt);
    await page.waitForChanges();

    expect(prevent).toHaveBeenCalled(); // navigation prevented
    expect(clickSpy).toHaveBeenCalled(); // still emits
  });
});
