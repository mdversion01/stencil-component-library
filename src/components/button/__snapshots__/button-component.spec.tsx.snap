// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`button-component (new spec) accordion props: sets aria-expanded/controls and toggles when isOpen changes 1`] = `
<button aria-controls="panel-1" aria-expanded="true" aria-label="Toggle Section" class="btn" type="button">
  <span class="btn__content">
    Toggle Section
  </span>
</button>
`;

exports[`button-component (new spec) applies ARIA attributes (aria-pressed, aria-label) on native button 1`] = `
<button aria-label="Submit form" aria-pressed="true" class="btn" type="button">
  <span class="btn__content">
    Submit
  </span>
</button>
`;

exports[`button-component (new spec) btnIcon (icon-only) renders slot directly, with aria-label fallback + dev warning 1`] = `
<button aria-label="Button" class="btn btn-icon" type="button">
  <span>
    üîî
  </span>
</button>
`;

exports[`button-component (new spec) grouped button with start placement adds group classes 1`] = `
<button aria-label="Start Btn" class="btn btn-group btn-group-start" type="button">
  <span class="btn__content">
    Start Btn
  </span>
</button>
`;

exports[`button-component (new spec) iconBtn wraps slot in .btn__content 1`] = `
<button aria-label="Button" class="btn icon-btn" type="button">
  <span class="btn__content">
    <span>
      üîî
    </span>
  </span>
</button>
`;

exports[`button-component (new spec) renders <a> link with aria-disabled and proper role/tabindex when disabled 1`] = `
<a aria-disabled="true" aria-label="Disabled Link" class="btn link" role="button" tabindex="-1">
  <span class="btn__content">
    Disabled Link
  </span>
</a>
`;

exports[`button-component (new spec) renders default <button> with text and stable aria-label 1`] = `
<button aria-label="Click me" class="btn" type="button">
  <span class="btn__content">
    Click me
  </span>
</button>
`;

exports[`button-component (new spec) slotSide="left" adds ms-1 and keeps me-1 off 1`] = `
<button aria-label="With Icon" class="btn" type="button">
  <span>
    üåü
  </span>
  <span class="btn__content ms-1">
    With Icon
  </span>
</button>
`;

exports[`button-component (new spec) slotSide="right" adds me-1 and keeps ms-1 off 1`] = `
<button aria-label="Next" class="btn" type="button">
  <span class="btn__content me-1">
    Next
  </span>
  <span>
    ‚û°Ô∏è
  </span>
</button>
`;
